src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,435 @@
+	 * way they can then be run in the same order. Tasks with the same
+	 * priority are served on a first-come-first-serve basis
+	 */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+		if (task->priority < task_iter->priority) {
+			list_item_append(&task->list, &task_iter->list);
+			break;
+		}
+	}
+
+	/*
+	 * If the task has not been added, it means that it has the lowest
+	 * priority and should be added at the end of the list
+	 */
+	if (list == &sch->tasks)
+		list_item_append(&task->list, &sch->tasks);
+}
+
+/*
+ * Task state machine:
+ * INIT:	initialized
+ * QUEUED:	inserted into the scheduler queue
+ * PENDING:	the scheduler is running, the task is moved to a temporary list
+ * RUNNING:	the scheduler is running, the task is being executed
+ * CANCEL:	anywhere, where the lock is not held, tasks can be cancelled
+ *		asynchronously from an arbitrary context
+ * other:	never set, shouldn't be used, but RESCHEDULE and COMPLETED are
+ *		returned by task's .run function, they are assigned to a
+ *		temporary state, but not to the task .state field
+ */
+
+static void zephyr_ll_run(void *data)
+{
+	struct zephyr_ll *sch = data;
+	struct task *task;
+	struct list_item *list, *tmp, head;
+	uint32_t flags;
+
+	list_init(&head);
+
+	spin_lock_irq(&sch->lock, flags);
+
+	/* Move all pending tasks to a temporary local list */
+	list_for_item_safe(list, tmp, &sch->tasks) {
+		struct comp_dev *sched_comp;
+		struct zephyr_ll_pdata *pdata;
+
+		task = container_of(list, struct task, list);
+		pdata = task->priv_data;
+
+		if (task->state == SOF_TASK_STATE_CANCEL) {
+			zephyr_ll_task_done(sch, task);
+			continue;
+		}
+
+		if (!domain_is_pending(sch->ll_domain, task, &sched_comp))
+			continue;
+
+		/*
+		 * Do not adjust .n_tasks, it's only decremented if the task is
+		 * removed via zephyr_ll_task_done()
+		 */
+		list_item_del(&task->list);
+		list_item_append(&task->list, &head);
+		pdata->run = true;
+
+		/*
+		 * In principle this is optional, it can also be overwritten if
+		 * the task is cancelled, which is an intrinsic race, that
+		 * cannot be prevented as long as we want to run tasks unlocked.
+		 */
+		task->state = SOF_TASK_STATE_PENDING;
+	}
+
+	spin_unlock_irq(&sch->lock, flags);
+
+	/*
+	 * Execute tasks on the temporary list. We are racing against
+	 * zephyr_ll_task_free(), see comments there for details.
+	 */
+	list_for_item_safe(list, tmp, &head) {
+		enum task_state state;
+		struct zephyr_ll_pdata *pdata;
+		bool must_run;
+
+		task = container_of(list, struct task, list);
+
+		spin_lock_irq(&sch->lock, flags);
+		if (task->state == SOF_TASK_STATE_PENDING) {
+			task->state = SOF_TASK_STATE_RUNNING;
+			must_run = true;
+		} else {
+			must_run = false;
+		}
+		spin_unlock_irq(&sch->lock, flags);
+
+		/*
+		 * While the lock is not held, the task can be cancelled, we
+		 * deal with it after the task completion.
+		 */
+		if (must_run) {
```

Comment:
why not just check task->state here ?

Reply:
having removed `PENDING` it can be done, yes